File: langs/python/src/vmx/capabilities/filter.py
"""Filter capability contract. See spec/14-capabilities.md §Filterable and ADR-0022."""

from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Callable
from typing import Generic, TypeVar

T = TypeVar("T")


class Filterable(ABC, Generic[T]):
    """A collection/composition that supports filtering by an arbitrary predicate.

    ``filter`` is the predicate; ``None`` means no filter applied.
    ``can_filter()`` reports whether filtering is currently allowed.
    """

    @property
    @abstractmethod
    def filter(self) -> Callable[[T], bool] | None: ...

    @filter.setter
    @abstractmethod
    def filter(self, value: Callable[[T], bool] | None) -> None: ...

    @abstractmethod
    def can_filter(self) -> bool: ...

File: langs/csharp/src/VMx/Capabilities/Filter.cs
namespace VMx.Capabilities;

// Capability interface from spec/14-capabilities.md §Filter and ADR-0022.
// Opt-in; not implemented by default by any core VM type.

/// <summary>Capability: the implementer accepts a typed predicate and exposes a filter action.</summary>
public interface IFilterable<TItem>
{
    /// <summary>The current filter predicate; <see langword="null"/> means no filter is applied.</summary>
    System.Predicate<TItem>? Filter { get; set; }

    /// <summary>Returns true when filtering is currently allowed.</summary>
    bool CanFilter();
}
